### ISSUE TITLE
(Bug report) Electron Port Can Collide

### ISSUE THREAD

**Issue by elementc** [zadam/trilium#3651]
### Trilium Version

0.58.8 (flatpak)

### What operating system are you using?

Ubuntu

### What is your setup?

Local + server sync

### Operating System Version

Linux WATERFALCON 5.19.0-29-generic #30-Ubuntu SMP PREEMPT_DYNAMIC Wed Jan 4 12:14:09 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Description

I've noticed that Trilium Desktop fails to start up sometimes.

I first noticed when clicking the launcher in my OS's desktop environment. I tried to start it from a terminal to see what was happening and saw that it was erroring out on a port collision, attached in the Error Logs section.

Indeed, the port it's picking, 37840, is in use by another service on my machine, `goa-daemon`:
```
casey on WATERFALCON in ~
at 12:32:15 🔥 lsof 2>/dev/null | grep 37840
goa-daemo    5165                              casey  179u     IPv4            2649091       0t0        TCP waterfalcon.lan:37840->REDACTED_DOMAIN:https (CLOSE_WAIT)
goa-daemo    5165    5168 gmain                casey  179u     IPv4            2649091       0t0        TCP waterfalcon.lan:37840->REDACTED_DOMAIN:https (CLOSE_WAIT)
goa-daemo    5165    5170 gdbus                casey  179u     IPv4            2649091       0t0        TCP waterfalcon.lan:37840->REDACTED_DOMAIN:https (CLOSE_WAIT)
goa-daemo    5165    5171 dconf\x20            casey  179u     IPv4            2649091       0t0        TCP waterfalcon.lan:37840->REDACTED_DOMAIN:https (CLOSE_WAIT)
```
`goa-daemon` is the GNOME Online Accounts Daemon. I don't want to disable it, nor should I.

This is using Trilium as a desktop app, they shouldn't use "well-known" ports, they should pick a free port and use that! This means a collision is resolved by a re-run, worst case.

@zadam [even said so once upon a time](https://github.com/zadam/trilium/issues/195#issuecomment-430919910):

> For desktop, random free port is used.

But every rerun sees the same error message, trying to use port 37840. This must be in the code somewhere! The offending line seems to be here:
https://github.com/zadam/trilium/blob/1b24276a4a2ab8e5b1814c8fe6ca222b9ab29de4/src/services/port.js#L22

Ok, we use a fixed port, and it can collide. We should fix that, to pick an actually free port, then use that, instead of using a well-known port. I can see in `port.js` that I may override this default port with the `TRILIUM_PORT` env variable. If you're here because this is happening to you, you can work around this issue right now with an invocation like this one, picking a different valid port number:
```
TRILIUM_PORT=37740 flatpak run com.github.zadam.trilium 
```
----
TL;DR of the bug:
* Is:
  * [This line of code uses an arbitrarily selected port which might already be in use.](https://github.com/zadam/trilium/blob/1b24276a4a2ab8e5b1814c8fe6ca222b9ab29de4/src/services/port.js#L22)
* Should be:
  * The code in question should find a free port and use that.
    * I don't know my way around node internals, but [these docs on the Node `net` module indicate that specifying port `0` will use an available port, if we're using `listen()` under the hood.](https://nodejs.org/docs/latest/api/net.html#serverlistenport-host-backlog-callback)
    * I attempted to test the theory of "use port `0` and we'll be fine", but port `0` seems to be [rejected by this parsing code.](https://github.com/zadam/trilium/blob/1b24276a4a2ab8e5b1814c8fe6ca222b9ab29de4/src/services/port.js#L9) This is as far as I can go in debugging this issue today.

### Error logs

```
casey on WATERFALCON in ~
at 12:09:52 🍀 flatpak run com.github.zadam.trilium
DB size: 13357 KB
Trusted reverse proxy: false
App HTTP server starting up at port 37840
{
  "appVersion": "0.58.8",
  "dbVersion": 212,
  "syncVersion": 29,
  "buildDate": "2023-02-13T21:50:54+01:00",
  "buildRevision": "17085e5578d2a20a77a6ade058f74e6d5b798ecc",
  "dataDirectory": "/home/casey/.local/share/trilium-data",
  "clipperProtocolVersion": "1.0",
  "utcDateTime": "2023-02-26T20:09:59.581Z"
}
CPU model: Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz, logical cores: 12 freq: 4352 Mhz
Port 37840 is already in use

casey on WATERFALCON in ~ 
at 12:10:00 🔥 
```

**Comment by zadam** [comment]
It used to be the case that Trilium used a random available port, but that changed with the addition of [Web Clipper](github.com/zadam/trilium-web-clipper) - the extension needs to know on which port Trilium is running.

Originally Trilium did that by picking a port in a small interval of 10 ports, but then there was a bug report that port scanning is extremely slow on some systems, so eventually this was resolved by using one fixed port.

**Comment by FliegendeWurst** [comment]
Linux uses ports 32768 to 60999 as epheremal ports: https://en.wikipedia.org/wiki/Ephemeral_port

To fix this issue (and #821) Trilium would need to listen on a port < 32768.

**Comment by LorenAmelang** [comment]
I just encountered this today. Looks like Thunderbird and Firefox (at least on Linux) both gradually walk through several of the "ephemeral ports" at a time, and will eventually grab 37480 for a session if Trilium doesn't have it first. Searching for Trilium instances won't find this problem, but  
`sudo netstat -peanut`
will show you
`tcp        0      0 10.1.1.9:37840          173.194.65.109:993      ESTABLISHED 1000       49710      6724/thunderbird    `
(That's Gmail it is connected to...)

Maybe the 
"Port 37840 is already in use. Most likely, another Trilium process is already running. You might try to find it, kill it, and try again" 
clue in the error report could mention that other apps might have grabbed it? 

Closing the Mozilla apps let Trilium start right up.

**Comment by eliandoran** [comment]
Closing in favor of #5298.
